File: Java/deepspace/src/deepspace/SpaceCity.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;

/**
 * Esta clase representa una ciudad espacial
 * @authors Manuel Diaz-Meco Terrés y Ana Graciani Donaire
 */
public class SpaceCity extends SpaceStation {
    
    // Atributos de instancia
    
    private ArrayList<SpaceStation> collaborators = null;
    private SpaceStation base = null;
    
    /**
     * Constructor con parámetros
     * @param b Base, estación espacial correspondiente al jugador actual
     * @param rest Colección de las estaciones espaciales del resto de jugadores
     */
    SpaceCity(SpaceStation b, ArrayList<SpaceStation> rest) {
        super(b);
        base = b;
        collaborators = new ArrayList<>(rest);
    }
    
    ArrayList<SpaceStation> getCollaborators() {
        return collaborators;
    }
    
    @Override
    public float fire() {
        float power = super.fire(); // base.fire()
        
        for (SpaceStation station : collaborators) {
            power+= station.fire();
        }
        
        return power;
    }
    
    @Override
    public float protection() {
        float shield = super.protection(); // base.protection()
        
        for (SpaceStation station : collaborators) {
            shield+= station.protection();
        }
        
        return shield;
    }
    
    @Override
    public Transformation setLoot(Loot loot) {
        super.setLoot(loot);
        return Transformation.NOTRANSFORM;
    }
    
    @Override
    public SpaceCityToUI getUIversion() {
        return new SpaceCityToUI(this);
    }
    
}

File: Ruby/SpaceCity.rb
#encoding: UTF-8

# Autores: Manuel Díaz-Meco Terrés y Ana Graciani Donaire
# 2ºDGIIM		PDOO

# Esta clase representa el concepto de ciudad espacial así como
# sus atributos y funcionalidades

require_relative 'SpaceStation'
require_relative 'SpaceCityToUI'

module Deepspace
	class SpaceCity < SpaceStation
		
		def initialize(base, rest)
			copy(base)
			@base = base
			@collaborators = rest
		end
		
		def collaborators
			return @collaborators
		end
		
		def fire
			power = super
			@collaborators.each do |s|
				power+= s.fire
			end
			return power
		end
		
		def protection
			shield = super
			@collaborators.each do |s|
				shield+= s.protection
			end
			return shield
		end
		
		def setLoot(loot)
			super
			return Transformation::NOTRANSFORM
		end
		
		def getUIVersion
			return SpaceCityToUI.new(self)
		end
		
		def to_s
			return getUIVersion.to_s
		end
			
	end
end
